### ISSUE TITLE
"no backups / backups visible / backups only" switch is undocumented.

### ISSUE THREAD

**Issue by kindaro** [DFHack/dfhack#1390]
It would be nice to have an in-game save backup management system. The presence of a relevant switch suggests it was at least partially implemented. However, there is no documentation to be found.

#### Try it yourself:

* Fire up `dfhack`.
* Choose "Continue Playing".
* Observe a switch in the lower right corner, associated with `b` key.

#### Expected:

I would like to see this switch described in the docs.

#### Actual:

I searched for "backup" in the docs, then Internet wide. Nothing relevant turned up.

* * *

I use Dwarf Fortress 0.44.12 and DFHack built from `develop`, commit `e56cb2a2`.

**Comment by kindaro** [comment]
This feature is possibly related to the obscure autosave option found in `data/init/d_init.txt`. If so, it should probably be disabled when said option is not enabled, to minimize confusion, and documented either way.

**Comment by lethosor** [comment]
It's part of gui/load-screen. Also, I think PyLNP has a control for the autosave and autobackup options.

**Comment by kindaro** [comment]
@lethosor I do not understand what you are saying. I am not aware of either `gui/load-screen` nor `PyLNP`.

**Comment by kindaro** [comment]
I enabled autosave in `data/init/d_init.txt` and I confirmed that the switch works with them as would be expected.

**Comment by kindaro** [comment]
Consider someone installing DF and DFHack outside of a starter pack. _(Which I did, for one.)_ This person may well not be aware of the "backup" feature of the game. _(That is me again.)_ Where do you expect such a person goes looking for information on this option?

There is nothing to make a person aware of the "backup" feature of the game before they meet the load screen and hence the switch in question. This feature is not even explicitly described in the [dwarf wiki page][1] corresponding to `d_init.txt` _(it does mention "autosave" but it is quite a shot from one term to another)_, and that would be quite obscure a resource to turn to. _(Where would a newcomer even get the idea of reading about this file from?)_

This is what the wiki says, for reference:

> The d_init.txt file can be used to make many adjustments to the game's behavior, such as enabling autosave, ...

[1]: http://dwarffortresswiki.org/index.php/DF2014:D_init.txt

This is literally all the documentation there is on gui/load-screen:

> A replacement for the “continue game” screen.
>
> Usage: `gui/load-screen enable|disable`

* * *

I am offering the argument above in support of two claims:

* The switch in question is in fact not self-explanatory, contrary to your opinion that is, possibly, biased due to you being a developer and thus, unlike a casual user, having extensive knowledge of most features of DF and DFHack, be they documented well or poorly.
* The very term "backup" is:
  - Not self-explanatory.
  - Not well-documented.
  - Not used consistently. _(Is it a "backup" or is it an "autosave"? That is, does the save get overwritten or a new one created every time?)_

I am hoping to display the facts at hand in such a way that we reach an agreement in that an amendment to the documentation is desirable.

**Comment by ab9rf** [comment]
I'm not sure that it's really incumbent on the DFHack implementers to make up for the lack of underlying feature documentation in Dwarf Fortress itself.

**Comment by lethosor** [comment]
Oh, I'm not expecting you to use a starter pack - I don't, and you're definitely not the only one that doesn't. I was just using that as an example of a fairly common thing that exposes the autosave/backup options.

The documentation in d_init.txt itself is fairly complete:
```
Use these to control the automatic saving behavior in the dwarf fortress mode of game.  AUTOSAVE can be set to NONE, SEASONAL or YEARLY.  This updates your save at these intervals, so that some of your progress will be saved in case of system instability.  You can set AUTOBACKUP to YES if you want the updated save to be copied to another folder so that you'll have several copies of your world at different times.  Set AUTOSAVE_PAUSE to YES if you want the game to pause every time it autosaves.

[AUTOSAVE:NONE]
[AUTOBACKUP:NO]
[AUTOSAVE_PAUSE:NO]
```
DF documents most settings in init.txt and d_init.txt in the files themselves, which is about the only useful place for it to put that documentation (besides the in-game help system, which isn't even searchable). The wiki is unofficial, and any missing information there is solely a wiki problem, not a DF/DFHack issue. It does point to d_init.txt for these types of settings, at any rate, and doesn't document all of them extensively.

I fail to see how the term "backup" isn't self-explanatory. Is there something else it could mean besides a copy of a save at a specific moment in time? That's what I use that term for in most instances outside of DF. Also, "autosaves" and "backups" aren't used inconsistently - they're two *different* features. An autosave is a save that is made automatically (without direct user action causing DF to save the game), while a backup is a copy of a save at a specific time that will not be overwritten. DF's d_init.txt explains this fairly well in my opinion.

Anyway, I see how "no backups" could be confusing if you thought it controlled the backups setting. Maybe "hide backups" would be better, and more consistent with the other two choices ("backups visible" and "backups only"). Perhaps one of those could be renamed "show backups", although that might be ambiguous (as one shows only backups and one includes them in the full list).

**Comment by kindaro** [comment]
I can make a further argument, that explains why I did not take the comments in `d_init.txt` in account.

It rests on the following statement:

Settings in `init.txt`, `d_init.txt` and other similar files are unsafe.

Therefore, I say these comments are aimed at developers / modders and do not constitute documentation for the casual user. I also question their searchability for a casual user: even if I see a search result like this in a general purpose Internet search engine, given without sufficient explication, I would disregard it because it looks technical and dangerous _(rightly so)_.

From here, I say if DFHack brings some of these unsafe and obscure options to the attention of the user, it is the responsibility of DFHack to provide documentation clarifying their use and purpose. If we put something in the user interface, we should at least:
* Make a claim that it is safe to use. _(Otherwise, why are we exposing a feature that is possibly unsafe?)_
* Offer a way to actually activate the feature. _(Otherwise, we are breeding confusion, and the feature cannot even be utilized by a measurable portion of users!)_

@ab9rf, do you disagree with any of these points?

Concerning the inconsistency of terms, I point out again that there is no documentation explicating the use of "backup" feature in a fashion that is accessible and friendly. What is mentioned in the wiki is the "autosave" feature. So, these two terms appear disconnected.

To clarify, I do not mean "backup" in general. I mean the term "backup" as it appears in the switch in question, where it refers to a _(supposed)_ specific feature of the game. In that sense, @lethosor would you still consider it self-explanatory? Can you explain how? Your explanation above applies only to the general meaning, which, I consider, would not be very useful for a person trying to actually benefit from the feature.

* * *

So, my argument goes this way:
* Some features in configuration files of DF are unsafe.
* Therefore, the casual user _(as opposed to modder or developer)_ should be expected to be aware of these features no more than they are to be aware of, say, memory layout.
* If we make use or mention of these features in the user interface, it is our responsibility to:
  - Ensure that these features are safe.
  - Explain how to make use of them.
  - Promise that it is safe to use them.
* Otherwise:
  - The user will not benefit from the feature.
  - They will be confused.
* Therefore, it is desirable to add some easily reachable and reliable documentation.

**Comment by ab9rf** [comment]
I fail to understand your assertion that _Settings in init.txt, d_init.txt and other similar files are unsafe._

Dwarf Fortress is, and remains, a work in progress. Everything in Dwarf Fortress is subject to change at any moment, if and when Toady has a reason to for it in the ineffable grand scheme of the development of his masterwork. Dwarf Fortress players understand this, or at least should, and therefore understand that the game is a jewel in the rough.

Basically, what I'm saying is that _there are no casual users of Dwarf Fortress_. Dwarf Fortress is, and will be for at least another decade, in a sort of public pre-alpha playtest. _Everything_ in Dwarf Fortress is "unsafe". But that's ok, because Losing Is Fun. 

Quite honestly, I don't think you understand the community that plays Dwarf Fortress.

**Comment by lethosor** [comment]
Yeah, I'm not sure why you think they're unsafe. I couldn't find any settings in d_init.txt that could break DF, and the few potentially buggy options in init.txt (e.g. PRINT_MODE settings) are mentioned. Did you see somewhere that said they were unsafe? Modifying files in raw/objects can certainly break saves if done wrong, but those are in a separate place.

The backup switch in gui/load-screen refers to saves created by the DF backup option. It offers an option to hide and show backup folders, which was requested because if they're all shown by default they take up a lot of space in the list. Would "hide backups" or "backups hidden" be more clear as to what it does?

I don't think hiding the option if AUTOBACKUP in d_init.txt is disabled is a good idea, because some people might only enable AUTOBACKUP sometimes. Not giving them an option to hide/show existing backups because they don't want to create new backups seems less useful. At any rate, I think the default of hiding backups is a reasonable default - it takes up less space, and people usually don't want to load backups directly. If they do, they'll know what option to look for.

**Comment by kindaro** [comment]
@ab9rf I see, we have a fundamental philosophical disagreement here.

* I stand on the position that we should strive to write safe and comfortable programs, including DF and DFHack.
* As I now may understand, your position is that DFHack should not aim to be comfortable, because it is for the elite few hardcore hackers.

You persuaded me. I am not going to play DF anymore.

@lethosor With your hard earned extensive knowledge and experience, you are confident that options in `d_init.txt`  are safe. There is no foundation to assume that confidence for a casual user. Any of these options may be broken in the most elaborate and unexpected way. After I set `VARIED_GROUND_TILES:NO`, gabbro stone under some of my underground stockpiles changed to above ground limestone.

I do think you are not trying to see things from the perspective of a person that has significantly less knowledge than you, on this highly particular subject.

Hiding the option is not a good idea. Now that I understand the way it is intended to work, I see that the best thing to do is to make the "backup" feature itself better known and more easily manageable.

My position, then, is that the "backup" feature is not discoverable. Your judgement on that is that the comments in `d_init.txt` are discoverable enough. Fine then. In any case, a sufficiently stubborn explorer will eventually stumble upon this very ticket, so at least some documentation we did create.

**Comment by lethosor** [comment]
I don't think @ab9rf meant that DFHack shouldn't be comfortable - we try to make DFHack user-friendly, at least (that's why gui/load-screen exists in the first place). She was mostly emphasizing that DF itself isn't comfortable to use, and we don't have any direct control over DF development.

> After I set `VARIED_GROUND_TILES:NO`, gabbro stone under some of my underground stockpiles changed to above ground limestone.

That's just a display option - it can't affect stone (at least in theory - I haven't bothered reverse-engineering anything to check). Gabbro appears to be a layer stone, but was it in clusters there? If you mine out clusters, the remaining floor turns into the base layer stone type as of 0.40, which could be limestone. That wouldn't alter the above-ground flag, though. Did you run any DFHack commands around the same time (like reveal, open-legends, etc.)?

I'm not ignoring the fact that players are unfamiliar with the autobackup option. I'm saying that the documentation for DFHack's gui/load-screen, which hardly even deals with backups, is not a good place to go into detail about DF's backup option. In general, documenting vanilla DF by itself is out of the scope of DFHack. There are certainly places where it could be better-documented, including the DF wiki.

I'm planning on making the label change, and can go a bit more in-depth in the gui/load-screen docs, so I'd like to keep this issue around for that, if you don't mind.

**Comment by kindaro** [comment]
@lethosor I think memory got corrupted and some bytes dropped to, say, 0x01, or whatever the code for sunlit limestone is. I observe it way below the sedimentary layer, and in exact squares. I do not actually think varied ground tiles are related, but my point is that I cannot know. The timing seems to correlate. Anyway, it only happened once. [The save, for reference.](https://github.com/DFHack/dfhack/files/2288133/region1.zip) See various stockpiles on levels 129, 130, 131.

I welcome and celebrate any improvements you see to be appropriate concerning this issue.

**Comment by lethosor** [comment]
Reopening since I forgot about this

**Comment by lethosor** [comment]
Hopefully addressed in dfhack/scripts@a242a7c. Sorry for the confusion and delay, and feel free to suggest improvements.
